### PR TITLE
Allow a trailing semicolon in block-closing-brace-*-after (#1806)

### LINKED ISSUE #1806: Allow a trailing semicolon after the closing brace of a block in block-closing-brace-*-after
CSS [custom property set](https://tabatkins.github.io/specs/css-apply-rule/#defining)  (used by `@apply` rules) is recognized as nested block due to its syntax, and so  triggers `block-closing-brace-newline-after` rule. 

This should be configurable by some extra option, e.g. `ignoreCustomPropertySets`.

> Which rule, if any, is this issue related to?

`block-closing-brace-newline-after`

> What CSS is needed to reproduce this issue?

```css
:root {
  --toolbar-theme: {
    background-color: hsl(120, 70%, 95%);
    border-radius: 4px;
    border: 1px solid var(--theme-color late);
  };
  --toolbar-title-theme: {
    color: green;
  };
}

.toolbar {
  @apply --toolbar-theme;
}
.toolbar > .title {
  @apply --toolbar-title-theme;
}

.warning {
  --toolbar-title-theme: {
    color: red;
    font-weight: bold;
  };
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "block-closing-brace-newline-before": "always-multi-line",
  }
}
```

> Which version of stylelint are you using?

`7.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to a [Tab Atkins draft](https://tabatkins.github.io/specs/css-apply-rule/)

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
 6:6   ✖  Expected newline after "}"
```

### PR DESCRIPTION
**Pull request by gaidarenko**
Closes #1806

### CODE REVIEW

File: src/rules/block-closing-brace-newline-after/README.md [hunk 1]
```diff
@@ -17,6 +17,14 @@ a {
 } /* end-of-line comment */
 ```
 
+This rule allows a trailing semicolon after the closing brace of a block. For example,
+
+```css
+a {
+  color: pink;
+};
```
Comment (evilebottnawi) [inline src/rules/block-closing-brace-newline-after/README.md]: This bad example, only for `:root` allow use trailing semicolon.

File: src/rules/block-closing-brace-newline-after/index.js [hunk 2]
```diff
@@ -62,17 +62,26 @@ export default function (expectation, options) {
...
       if (!nodeToCheck) { return }
 
+      let reportIndex = statement.toString().length
+      let src = rawNodeString(nodeToCheck)
+
+      // Skip a semicolon at the beginning, if any
+      if (src && src[0] === ";") {
+        src = src.slice(1)
+        reportIndex++
+      }
```
Comment (evilebottnawi) [inline src/rules/block-closing-brace-newline-after/index.js]: See above, i don't see check on `:root`

File: src/rules/block-closing-brace-space-after/README.md [hunk 3]
```diff
@@ -8,6 +8,12 @@ a { color: pink; }
  * The space after this brace */
 ```
 
+This rule allows a trailing semicolon after the closing brace of a block. For example,
+
+```css
+a { color: pink; };
```
Comment (evilebottnawi) [inline src/rules/block-closing-brace-space-after/README.md]: As above

File: src/rules/block-closing-brace-space-after/index.js [hunk 4]
```diff
@@ -46,15 +46,24 @@ export default function (expectation) {
...
 
+      let reportIndex = statement.toString().length
+      let src = rawNodeString(nextNode)
+
+      // Skip a semicolon at the beginning, if any
+      if (src && src[0] === ";") {
+        src = src.slice(1)
+        reportIndex++
+      }
+
```
Comment (evilebottnawi) [inline src/rules/block-closing-brace-space-after/index.js]: As above

**Comment by evilebottnawi** [comment]
@gaidarenko need add check on only `:root`

**Comment by gaidarenko** [comment]
In [this discussion](https://github.com/stylelint/stylelint/issues/1806)  @davidtheclark and @jeddy3 were talking only about this:

> we could just allow for a ; before the expected newline, right?

As far as I understand they propose to allow for `;` in any declaration without restrictions. Am I wrong?

**Comment by evilebottnawi** [comment]
@gaidarenko let's wait @davidtheclark and @jeddy3 says about this, in CSS valid only for `:root` trailing semicolon, also check only on `:root` speedup rule

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,6 @@
 # Head
 
+-   Fixed: `block-closing-brace-newline-after` and `block-closing-brace-space-after` now allows a trailing semicolon after the closing brace of a block.
```
Comment (davidtheclark) [inline CHANGELOG.md]: allows -> allow

File: src/rules/block-closing-brace-newline-after/README.md (hunk 1 above)
Comment (davidtheclark) [inline src/rules/block-closing-brace-newline-after/README.md]: Yep. Here's what I think's a good example from the original issue:

```css
:root {
  --toolbar-theme: {
    background-color: hsl(120, 70%, 95%);
    border-radius: 4px;
    border: 1px solid var(--theme-color late);
  };
}
```

**Comment by davidtheclark** [comment]
@evilebottnawi In https://tabatkins.github.io/specs/css-apply-rule/#defining it doesn't look to me like that syntax is only valid on `:root`. Example 2 shows:

```css
.foo {
  --my-theme: {
    color: red;
    background: white;
  };
}
```

That makes me think the examples should focus on `:root` but the code and tests should look outside of `:root`, as well.

**Comment by evilebottnawi** [comment]
@davidtheclark sgtm

**Comment by gaidarenko** [comment]
I have adjusted the code according to your comments.

**Comment by jeddy3** [comment]
LGTM

@gaidarenko Thanks for another nice fix! :)

File: src/rules/block-closing-brace-newline-after/index.js [hunk 6]
```diff
@@ -62,17 +62,26 @@ export default function (expectation, options) {
       const nodeToCheck = (nextNodeIsSingleLineComment) ? nextNode.next() : nextNode
       if (!nodeToCheck) { return }
 
+      let reportIndex = statement.toString().length
+      let src = rawNodeString(nodeToCheck)
```
Comment (davidtheclark) [inline src/rules/block-closing-brace-newline-after/index.js]: Let's just name this variable `source`: no need to abbreviate.

File: src/rules/block-closing-brace-space-after/README.md [hunk 7]
```diff
@@ -8,6 +8,12 @@ a { color: pink; }
  * The space after this brace */
 ```
 
+This rule allows a trailing semicolon after the closing brace of a block. For example,
+
+```css
+:root { --toolbar-theme: { background-color: hsl(120, 70%, 95%); }; }
```
Comment (davidtheclark) [inline src/rules/block-closing-brace-space-after/README.md]: Please make this example multi-line. It will be easier to read.

**Comment by davidtheclark** [comment]
@gaidarenko As discussed above, I think we need some tests that ensure this works for blocks outside `:root`.

**Comment by gaidarenko** [comment]
@davidtheclark Done.

**Comment by davidtheclark** [comment]
Great, thanks @gaidarenko!

**Pull request merged**